applications/rag/frontend/container/rag_langchain/rag_chain.py

```diff
@@ -0,0 +1,132 @@
+History: {""" + HISTORY + "}\n\nContext: {" + CONTEXT + "}\n\nQuestion: {" + QUESTION + "}\n"
+
+prompt = PromptTemplate(template=template, input_variables=[HISTORY, CONTEXT, QUESTION])
+
+engine = create_sync_postgres_engine()
+# TODO: Dict is not safe for multiprocessing. Introduce a cache using Flask-caching or libcache
+# The in-memory SimpleCache implementations for each of these libraries is not safe either. 
+# Consider redis or memcached (e.g., Memorystore)
+# chat_history_map: Dict[str, PostgresChatMessageHistory] = {}
+
+def get_chat_history(session_id: str) -> PostgresChatMessageHistory:
+    history = PostgresChatMessageHistory.create_sync(
+        engine,
+        session_id=session_id,
+        table_name = CHAT_HISTORY_TABLE_NAME
+    )
+    # history = None
+    # if session_id in chat_history_map:
+    #     history = chat_history_map[session_id]
+    # else:
+    #     history = PostgresChatMessageHistory.create_sync(
+    #         engine,
+    #         session_id=session_id,
+    #         table_name = CHAT_HISTORY_TABLE_NAME
+    #     )
+    #     chat_history_map[session_id] = history
+    print(f"Retrieving history for session {session_id} with {len(history.messages)} items:")
+    print("\t" + "\n\t".join([message.content for message in history.messages]))
+    return history
+
+def clear_chat_history(session_id: str):
+    history = PostgresChatMessageHistory.create_sync(
+        engine,
+        session_id=session_id,
+        table_name = CHAT_HISTORY_TABLE_NAME
+    )    # if session_id in chat_history_map:
+    #     history = chat_history_map[session_id]
+    history.clear()
+        # del chat_history_map[session_id]
+
+#TODO: limit number of tokens in prompt to MAX_INPUT_LENGTH 
+# (as specified in hugging face TGI input parameter)
+
+def create_chain() -> RunnableWithMessageHistory: 
+    # TODO HuggingFaceTextGenInference class is deprecated. 
+    # The warning is: 
+    #                The class `langchain_community.llms.huggingface_text_gen_inference.HuggingFaceTextGenInference` 
+    #                was deprecated in langchain-community 0.0.21 and will be removed in 0.2.0. Use HuggingFaceEndpoint instead
+    # The replacement is HuggingFace Endoint, which requires a huggingface
+    # hub API token. Either need to add the token to the environment, or need to find a method to call TGI
+    # without the token.
+    # Example usage of HuggingFaceEndpoint:
+    # llm = HuggingFaceEndpoint(
+    #                 endpoint_url=f'http://{INFERENCE_ENDPOINT}/',
+    #                 max_new_tokens=512,
+    #                 top_k=10,
+    #                 top_p=0.95,
+    #                 typical_p=0.95,
+    #                 temperature=0.01,
+    #                 repetition_penalty=1.03,
+    #                 huggingfacehub_api_token="my-api-key"
+    #             )
+    model = HuggingFaceTextGenInference(
+        inference_server_url=f'http://{INFERENCE_ENDPOINT}/',
+        max_new_tokens=512,
```

Comment:
can we override max_new_tokens here instead? or did that not work

Reply:
This is a different parameter than `MAX_INPUT_LENGTH` and `MAX_TOTAL_TOKENS`. Changing it does not have the desired effect.